core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+
+            // TODO implement small tasks merging
+
+            int taskPartitionId = 0;
+            ImmutableList.Builder<TaskDescriptor> result = ImmutableList.builder();
+            for (Integer partition : union(partitionToSplitsMap.keySet(), partitionToExchangeSourceHandlesMap.keySet())) {
+                Multimap<PlanNodeId, Split> splits = partitionToSplitsMap.getOrDefault(partition, ImmutableMultimap.of());
+                Multimap<PlanNodeId, ExchangeSourceHandle> exchangeSourceHandles = ImmutableListMultimap.<PlanNodeId, ExchangeSourceHandle>builder()
+                        .putAll(partitionToExchangeSourceHandlesMap.getOrDefault(partition, ImmutableMultimap.of()))
+                        .putAll(replicatedExchangeSourceHandles)
+                        .build();
+                Optional<Set<HostAddress>> hostRequirement = Optional.ofNullable(partitionToNodeMap.get(partition)).map(ImmutableSet::of);
+                result.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement)));
+            }
+
+            finished = true;
+            return result.build();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return finished;
+        }
+
+        @Override
+        public void close()
+        {
+            if (closed) {
+                return;
+            }
+            closed = true;
+            for (SplitSource splitSource : splitSources.values()) {
+                try {
+                    splitSource.close();
+                }
+                catch (RuntimeException e) {
+                    log.error(e, "Error closing split source");
+                }
+            }
+        }
+    }
+
+    public static class SourceDistributionTaskSource
+            implements TaskSource
+    {
+        private final QueryId queryId;
+        private final PlanNodeId partitionedSourceNodeId;
+        private final TableExecuteContextManager tableExecuteContextManager;
+        private final SplitSource splitSource;
+        private final Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles;
+        private final int splitBatchSize;
+        private final LongConsumer getSplitTimeRecorder;
+        private final Optional<CatalogName> catalogRequirement;
+        private final int targetPartitionSplitCount;
+
+        private final Queue<Split> remotelyAccessibleSplitBuffer = new ArrayDeque<>();
+        private final Map<HostAddress, Set<Split>> locallyAccessibleSplitBuffer = new HashMap<>();
+
+        private int currentPartitionId;
+        private boolean finished;
+        private boolean closed;
+
+        public static SourceDistributionTaskSource create(
+                Session session,
+                PlanFragment fragment,
+                Multimap<PlanFragmentId, ExchangeSourceHandle> exchangeSourceHandles,
+                SplitSourceFactory splitSourceFactory,
```

Comment:
nit: swap `splitSourceFactory` and `exchangeSourceHandles` to be consistent with `HashDistributionTaskSource.create`